### PR TITLE
exit codes, logger and password to typescript

### PR DESCRIPTION
**Pull request by foxriver76**

### CODE REVIEW

File: packages/common/src/lib/common/exitCodes.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-const EXIT_CODES = {
+const EXIT_CODES: Record<string | number, string | number> = {
```
Comment (AlCalzone) [inline packages/common/src/lib/common/exitCodes.ts:3]: This should be an `enum` instead, so we don't lose the syntax help for availabl properties.
`enum`s also include the reverse lookup, so that is no longer necessary.

File: packages/common/src/lib/common/password.ts [hunk 2]
```diff
@@ -29,27 +29,28 @@
...
-let version = null;
-
-const password = pw => {
+export const password = (
+    pw: string
+): {
+    complexity: (password: string, callback: (isComplex: boolean) => void) => boolean;
+    check: (hashedPassword: string, callback: (err?: Error | null, isOk?: boolean) => void) => void;
+    hash: (salt: string, iterations: number, callback: (err?: Error | null, hash?: string) => void) => void;
+} => {
```
Comment (AlCalzone) [inline packages/common/src/lib/common/password.ts:41]: The return type is complex enough to factor out into an `interface`.
Also, `@types/iobroker` has a helper type for our common error-result-callbacks (`GenericCallback<returntype>`) that you could use here.

File: packages/common/src/lib/common/password.ts [hunk 3]
```diff
@@ -67,16 +68,16 @@ const password = pw => {
                 });
             }
         },
-        check: function (hashedPassword, callback) {
+        check: function (hashedPassword: string, callback: (err?: Error | null, isOk?: boolean) => void) {
```
Comment (AlCalzone) [inline packages/common/src/lib/common/password.ts:71]: It shouldn't be necessary to repeat the function type annotations here, since they are already on the method return type.

File: packages/common/src/lib/common/password.ts [hunk 4]
```diff
@@ -29,27 +29,28 @@
...
             iterations = iterations || 10000;
 
             // version 0.10 has no 'sha256' and this option must be ignored
-            if (version === null) {
-                version = process.version.replace('v', '');
-                version = !version.startsWith('0.10.');
+            if (versionOk === undefined) {
+                const version = process.version.replace('v', '');
+                versionOk = !version.startsWith('0.10.');
             }
```
Comment (AlCalzone) [inline packages/common/src/lib/common/password.ts:51]: Do we really still need to support node 0.10? 🤔

File: packages/common/src/lib/common/password.ts [hunk 5]
```diff
@@ -29,27 +29,28 @@
  *
  */
 
-/* jshint -W097 */
-/* jshint strict: false */
-/* jslint node: true */
-'use strict';
+import * as crypto from 'crypto';
```
Comment (AlCalzone) [inline packages/common/src/lib/common/password.ts:32]: With `esModuleInterop: true` in the root tsconfig (I think it is in @tsconfigs/node), you can shorten this to 
```suggestion
import crypto from 'crypto';
```

**Review by AlCalzone — CHANGES_REQUESTED** [review]

File: packages/common/src/lib/common/exitCodes.ts (hunk 1 above)
Comment (AlCalzone) [reply packages/common/src/lib/common/exitCodes.ts:3]: (This is the one TypeScript feature that has no equivalent in JS, but that I'd always miss if it were gone)

File: packages/common/src/lib/common/tools.js [hunk 6]
```diff
@@ -475,7 +475,7 @@ async function createUuid(objects) {
     const promiseCheckPassword = new Promise(resolve =>
         objects.getObject('system.user.admin', (err, obj) => {
             if (err || !obj) {
-                password = password || require('./password');
+                password = password || require('./password').password;
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.js:478]: Any reason we can't statically import `./password` at the top?

Comment (foxriver76) [reply packages/common/src/lib/common/tools.js]: bf style

File: packages/common/src/lib/common/exitCodes.ts (hunk 1 above)
Comment (foxriver76) [reply packages/common/src/lib/common/exitCodes.ts:3]: does it work if you take the reverse lookup into account?

Comment (foxriver76) [reply packages/common/src/lib/common/exitCodes.ts:3]: doesnt seem so, because as soon as i define it as an enum it will be read-only

File: packages/common/src/lib/common/tools.ts [hunk 7]
```diff
@@ -3212,7 +3212,7 @@ function removePreservedProperties(
  * @param namespace - adapter namespace + id, e.g. hm-rpc.0
  * @param createWakeup - indicator to create wakeup object too
  */
-function getInstanceIndicatorObjects(namespace: string, createWakeup: boolean): any[] {
+function getInstanceIndicatorObjects(namespace: string, createWakeup: boolean): Record<string, any>[] {
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts]: We have proper types for every object type, I think this is the correct one:
```suggestion
function getInstanceIndicatorObjects(namespace: string, createWakeup: boolean): Record<string, ioBroker.StateObject>[] {
```

Comment (foxriver76) [reply packages/common/src/lib/common/tools.ts]: yes, strange it already looks like this for a few commits

File: lerna-debug.log [hunk 8]
```diff
@@ -0,0 +1,21 @@
+0 silly argv {
```
Comment (AlCalzone) [inline lerna-debug.log:1]: I don't think this was supposed to be checked in

File: packages/common/src/lib/common/exitCodes.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
+'use strict';
```
Comment (AlCalzone) [inline packages/common/src/lib/common/exitCodes.ts:1]: You can remove this in all TS files.

File: packages/common/src/lib/common/logger.ts [hunk 10]
```diff
@@ -125,24 +152,24 @@ const logger = function (level, files, noStdout, prefix) {
...
     files = files || [];
 
     // indicator which is used to determine the log dir for developing, where it should be inside the repository
     const isNpm = !__dirname
         .replace(/\\/g, '/')
         .toLowerCase()
-        .includes(tools.appName.toLowerCase() + `.js-controller/packages/`);
+        .includes(`${tools.appName.toLowerCase()}.js-controller/packages/`);
 
     if (typeof level === 'object') {
```
Comment (AlCalzone) [inline packages/common/src/lib/common/logger.ts:165]: Better use `isObject` helper. Even if we limit the type to something non-nullable, seeing `typeof ... === "object"` checks makes me nervous :)

File: packages/common/src/lib/common/logger.ts [hunk 11]
```diff
@@ -180,7 +207,7 @@ const logger = function (level, files, noStdout, prefix) {
                     }
 
                     transport.handleExceptions = false;
-                    transport.name = !fName ? tools.appName : 'dailyRotateFile' + fName;
+                    transport.name = !fName ? tools.appName : `dailyRotateFile${fName}`;
```
Comment (AlCalzone) [inline packages/common/src/lib/common/logger.ts:210]: Just personal preference, I'd turn the ternary around.

File: packages/common/src/lib/common/logger.ts [hunk 12]
```diff
@@ -113,9 +120,29 @@ class NotifierTransport extends Transport {
...
+    prefix?: string;
+    transport: Record<string, any>;
+}
+
+interface Options {
+    format?: winston.Logform.Format;
+    transports: Transport[];
+}
+
+export const logger = function (
```
Comment (AlCalzone) [inline packages/common/src/lib/common/logger.ts:139]: is there a reason that this is not `export function logger(...)`?

File: packages/common/src/lib/common/password.ts [hunk 13]
```diff
@@ -29,54 +29,42 @@
...
+    check: (hashedPassword: string, callback: (err?: Error | null, isOk?: boolean) => void) => void;
+    hash: (
+        salt: string | null,
+        iterations: number | null,
+        callback: (err?: Error | null, hash?: string) => void
+    ) => void;
+}
 
-const password = pw => {
+export const password = (pw: string): PasswordReturnValue => {
```
Comment (AlCalzone) [inline packages/common/src/lib/common/password.ts:44]: Should be `export function password(...)`

File: packages/common/src/lib/common/password.ts [hunk 14]
```diff
@@ -90,16 +78,15 @@ const password = pw => {
         complexity: (password, callback) => {
             let result = false;
             if (typeof password === 'string') {
-                result =
+                result = !!(
                     password.length >= 8 && // minimum length is 8
                     password.match(/\d/) && // contains at least one digit
                     password.match(/[a-z]/) && // contains at least one lower case letter
-                    password.match(/[A-Z]/); // contains at least one upper case letter
+                    password.match(/[A-Z]/)
```
Comment (AlCalzone) [inline packages/common/src/lib/common/password.ts:85]: These checks should be `/regex/.test(string)`, then you can drop the `!!`

File: packages/common/src/lib/common/password.ts [hunk 15]
```diff
@@ -90,16 +78,15 @@ const password = pw => {
...
+                ); // contains at least one upper case letter
             }
             typeof callback === 'function' && callback(result);
```
Comment (AlCalzone) [inline packages/common/src/lib/common/password.ts:88]: This can be written more concise as `callback?.(result)`, assuming callback is never something that's not a function or nullable.

**Review by AlCalzone — CHANGES_REQUESTED** [review]
First part of the review - I can't open tools.ts locally, so this will come in a second round.

File: packages/common/src/lib/common/logger.ts (hunk 10 above)
Comment (foxriver76) [reply packages/common/src/lib/common/logger.ts:165]: yes now that tools is ts I can do that, else he wouldn't get that it is of type object after the check :+1:

File: packages/common/src/lib/common/tools.ts [hunk 16]
```diff
@@ -839,24 +859,29 @@ function scanDirectory(dirName, list, regExp) {
...
 /**
  * Get list of all installed adapters and controller version on this host
- * @param {string} [hostRunningVersion] Version of the running js-controller, will be included in the returned information if provided
- * @returns {object} object containing information about installed host
+ * @param hostRunningVersion Version of the running js-controller, will be included in the returned information if provided
+ * @returns object containing information about installed host
  */
-function getInstalledInfo(hostRunningVersion) {
-    const result = {};
+function getInstalledInfo(hostRunningVersion?: string): Record<string, any> {
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts:863]: We should define a proper return type, so you know from outside what you get.

File: packages/common/src/lib/common/tools.ts [hunk 17]
```diff
@@ -1142,24 +1176,31 @@ function _checkRepositoryFileHash(urlOrPath, additionalInfo, callback) {
...
+ * @param callback function (err, sources, actualHash) { }
  *
  */
-function getRepositoryFile(urlOrPath, additionalInfo, callback) {
-    let sources = {};
+function getRepositoryFile(
+    urlOrPath: string,
+    additionalInfo: Record<string, any>,
+    callback: (
+        err?: undefined | Error | null,
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts:1186]: `?:` automatically includes `undefined` in the type.

File: packages/common/src/lib/common/tools.ts [hunk 18]
```diff
@@ -1250,16 +1291,20 @@ function getRepositoryFile(urlOrPath, additionalInfo, callback) {
...
+ * @param _actualRepo Actual repository
  *
  */
-async function getRepositoryFileAsync(url, hash, force, _actualRepo) {
+async function getRepositoryFileAsync(
+    url: string,
+    hash: string,
+    force: boolean,
+    _actualRepo: Record<string, any>
+): Promise<Record<string, any>> {
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts:1303]: Probably should define how the repo object looks like

File: packages/common/src/lib/common/tools.ts [hunk 19]
```diff
@@ -1250,16 +1291,20 @@ function getRepositoryFile(urlOrPath, additionalInfo, callback) {
...
     let _hash;
     if (_actualRepo && !force && hash && (url.startsWith('http://') || url.startsWith('https://'))) {
-        axios = axios || require('axios');
         _hash = await axios({ url: url.replace(/\.json$/, '-hash.json'), timeout: 10000 });
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts:1306]: We're using `axios` here, but `request` elsewhere. We should get rid of `request`.

File: packages/common/src/lib/common/tools.ts [hunk 20]
```diff
@@ -1269,16 +1314,16 @@ async function getRepositoryFileAsync(url, hash, force, _actualRepo) {
...
         }
 
         if (_actualRepo && hash && _hash && _hash.data && _hash.data.hash === hash) {
             data = _actualRepo;
         } else {
-            const agent = `${module.exports.appName}, RND: ${randomID}, Node:${process.version}, V:${
-                require('../../package.json').version
+            const agent = `${exports.default.appName}, RND: ${randomID}, Node:${process.version}, V:${
+                // eslint-disable-next-line @typescript-eslint/no-var-requires
+                require('../../../package.json').version
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts:1324]: We should require the actual module name `@iobroker/js-controller-common/package.json` instead of relying on relative paths.

File: packages/common/src/lib/common/tools.ts [hunk 21]
```diff
@@ -1309,13 +1354,11 @@ async function getRepositoryFileAsync(url, hash, force, _actualRepo) {
 /**
  * Sends the given object to the diagnosis server
  *
- * @param {object} obj - diagnosis object
- * @return {Promise<void>}
+ * @param obj - diagnosis object
  */
-async function sendDiagInfo(obj) {
+async function sendDiagInfo(obj: Record<string, any>): Promise<void> {
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts:1357]: Also we should define how the object actually looks like.

File: packages/common/src/lib/common/tools.ts [hunk 22]
```diff
@@ -1382,19 +1425,18 @@ function getAdapterDir(adapter) {
...
+function getHostName(): string {
     // for tests purposes
     if (process.env.IOB_HOSTNAME) {
         return process.env.IOB_HOSTNAME;
     }
     try {
         const configName = getConfigFileName();
         const config = fs.readJSONSync(configName);
-        return config.system ? config.system.hostname || require('os').hostname() : require('os').hostname();
+        return config.system ? config.system.hostname || os.hostname() : os.hostname();
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts]:
```suggestion
        return config.system?.hostname || os.hostname();
```
(or `??` instead of `||`) if we know that hostname is never empty.

File: packages/common/src/lib/common/tools.ts [hunk 23]
```diff
@@ -1917,13 +1971,13 @@ async function getHostInfo(objects, callback) {
         data.Platform = 'OSX';
     }
 
-    const systemConfig = await objects.getObjectAsync('system.config');
-    const systemRepos = await objects.getObjectAsync('system.repositories');
+    const systemConfig: ioBroker.Object = await objects.getObjectAsync('system.config');
+    const systemRepos: ioBroker.Object = await objects.getObjectAsync('system.repositories');
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts]: Once the objects DB has proper types, we should remove these annotations so the objects get their actual type that's inferred from the IDs.

File: packages/common/src/lib/common/tools.ts [hunk 24]
```diff
@@ -2047,15 +2098,15 @@ function getConfigFileName() {
 
     // if debugging with npm5 -> node_modules on e.g. /opt/node_modules
     if (
-        fs.existsSync(`${__dirname}/../../../../../../../node_modules/${appName.toLowerCase()}.js-controller`) ||
-        fs.existsSync(`${__dirname}/../../../../../../../node_modules/${appName}.js-controller`)
+        fs.existsSync(`${__dirname}/../../../../../../../../node_modules/${appName.toLowerCase()}.js-controller`) ||
+        fs.existsSync(`${__dirname}/../../../../../../../../node_modules/${appName}.js-controller`)
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts]: 🤮

File: packages/common/src/lib/common/tools.ts [hunk 25]
```diff
@@ -2080,23 +2131,27 @@ function sliceArgs(argsObj, startIndex) {
...
- * @param {string[]} [returnArgNames] (optional) If the callback contains multiple arguments,
+ * @param fn The function to promisify
+ * @param context (optional) The context (value of `this` to bind the function to)
+ * @param returnArgNames (optional) If the callback contains multiple arguments,
  * you can combine them into one object by passing the names as an array.
  * Otherwise the Promise will resolve with an array
- * @returns {(...args: any[]) => Promise<any>}
  */
-function promisify(fn, context, returnArgNames) {
+function promisify(
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts]: We should kill these functions at some point. I think Node.js is far enough now that we don't need them anymore.

File: packages/common/src/lib/common/tools.ts [hunk 26]
```diff
@@ -2184,44 +2243,40 @@ function promisifyNoError(fn, context, returnArgNames) {
...
+            // ignore
+        }
     }
 }
 
-function setQualityForInstance(objects, states, namespace, q) {
-    return new Promise((resolve, reject) => {
+function setQualityForInstance(objects: any, states: any, namespace: string, q: number): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
         objects.getObjectView(
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts]: Can be made async I think.

File: packages/common/src/lib/common/tools.ts [hunk 27]
```diff
@@ -2359,20 +2407,18 @@ function encrypt(key, value) {
...
         return decryptLegacy(key, value);
     }
 
-    crypto = crypto || require('crypto');
-
-    const textParts = value.split(':');
+    const textParts: string[] = value.split(':');
     const iv = Buffer.from(textParts[1], 'hex');
+    /** @ts-expect-error ts doesn't know if an element is in */
     const encryptedText = Buffer.from(textParts.pop(), 'hex');
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts]: I think this captures the intent better (if I didn't misunderstand) and also documents what the parts are.
```suggestion
    const textParts = value.split(':', 3) as [alg: string, iv: string, plaintext: string];
    const iv = Buffer.from(textParts[1], 'hex');
    const encryptedText = Buffer.from(textParts[2], 'hex');
```

File: packages/common/src/lib/common/tools.ts [hunk 28]
```diff
@@ -2615,11 +2667,13 @@ async function removeIdFromAllEnums(objects, id, allEnums) {
...
  * @memberof tools
- * @param {string[]|Record<string, string>[]|string|Record<string, string>} dependencies dependencies array or single dependency
- * @returns {Record<string, string>} parsed dependencies
+ * @param dependencies dependencies array or single dependency
+ * @returns parsedDeps parsed dependencies
  */
-function parseDependencies(dependencies) {
-    let adapters = {};
+function parseDependencies(
+    dependencies: string[] | Record<string, string>[] | string | Record<string, string>
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts]: Gotta love a clean and specific API 🙈

File: packages/common/src/lib/common/tools.ts [hunk 29]
```diff
@@ -3378,32 +3456,42 @@ function getLogger(log) {
     return log;
 }
 
+interface OrderedInstancesObject {
+    1: ioBroker.InstanceObject[];
+    2: ioBroker.InstanceObject[];
+    3: ioBroker.InstanceObject[];
+    admin: ioBroker.InstanceObject[];
+}
```
Comment (AlCalzone) [inline packages/common/src/lib/common/tools.ts]: Could also be written as
```ts
type OrderedInstancesObject = Record<"1" | "2" | "3" | "admin", ioBroker.InstanceObject[]>;
```

**Review by AlCalzone — CHANGES_REQUESTED** [review]
Probably missed a bunch, but here we gooooo

File: packages/common/src/lib/common/tools.ts (hunk 28 above)
Comment (foxriver76) [reply packages/common/src/lib/common/tools.ts]: historically these are the dependencies which were allowed in io-pack at some point

File: packages/common/src/lib/common/tools.ts (hunk 26 above)
Comment (foxriver76) [reply packages/common/src/lib/common/tools.ts]: yes it can but not necessarily in this PR :smile:

File: packages/common/src/lib/common/tools.ts [hunk 30]
```diff
@@ -1,33 +1,42 @@
...
+const extend = require('node.extend');
+
+const ERROR_NOT_FOUND = 'Not exists';
+const ERROR_EMPTY_OBJECT = 'null object';
+const ERROR_NO_OBJECT = 'no object';
+const ERROR_DB_CLOSED = 'DB closed';
+
+events.EventEmitter.prototype.setMaxListeners(100);
+let npmVersion: string;
+let diskusage: typeof import('diskusage');
```
[the remark this answers is not in the document]
Comment (foxriver76) [reply packages/common/src/lib/common/tools.ts:32]: I think diskusage is an optional dep thus I went for the dynamic import

File: packages/common/src/lib/common/tools.ts [hunk 31]
```diff
@@ -198,47 +209,62 @@ function decryptPhrase(password, data, callback) {
...
+ */
+async function isHostRunning(objects: any, states: any): Promise<boolean> {
+    const res: GetObjectViewResult = await objects.getObjectViewAsync('system', 'host', {
+        startkey: '',
+        endkey: '\u9999'
+    });
 
     for (const hostObj of res.rows) {
-        const state = await states.getState(`${hostObj.id}.alive`);
+        const state: ioBroker.State = await states.getState(`${hostObj.id}.alive`);
```
[the remark this answers is not in the document]
Comment (foxriver76) [reply packages/common/src/lib/common/tools.ts:236]: yes, one step at a time ;-)

File: packages/common/src/lib/common/tools.ts [hunk 32]
```diff
@@ -288,20 +311,20 @@ function findPath(path, url) {
             return (path + url).replace(/\/\//g, '/').replace('http:/', 'http://').replace('https:/', 'https://');
```
[the remark this answers is not in the document]
Comment (foxriver76) [reply packages/common/src/lib/common/tools.ts:309]: dafuq is this

File: packages/common/src/lib/common/tools.ts [hunk 33]
```diff
@@ -468,17 +494,15 @@ function updateUuid(newUuid, _objects, callback) {
...
-        objects.getObject('system.user.admin', (err, obj) => {
+async function createUuid(objects: any): Promise<void | string> {
+    const promiseCheckPassword = new Promise<void>(resolve =>
+        objects.getObject('system.user.admin', (err: Error | null, obj: ioBroker.UserObject) => {
             if (err || !obj) {
-                password = password || require('./password');
-
                 // Default Password for user 'admin' is application name in lower case
-                password(module.exports.appName).hash(null, null, (err, res) => {
+                password(exports.default.appName).hash(null, null, (err, res) => {
```
[the remark this answers is not in the document]
Comment (foxriver76) [reply packages/common/src/lib/common/tools.ts:491]: I dont really get it.. so exporting every function on its own? And your cited code in index.ts?

[the remark this answers is not in the document]
Comment (AlCalzone) [reply packages/common/src/lib/common/tools.ts:491]: Nevermind, this doesn't really make sense. The exports from `tools` used to be named exports, so it really shouldn't be necessary to make a fallback default export.
See my description in the chat for context.

File: packages/common/src/lib/common/tools.ts (hunk 25 above)
Comment (foxriver76) [reply packages/common/src/lib/common/tools.ts:2126]: We use them in adapter.js a lot currently

File: packages/common/src/lib/common/tools.ts (hunk 24 above)
Comment (foxriver76) [reply packages/common/src/lib/common/tools.ts:2088]: if debugging with npm 5.. not sure if this is still needed?

File: packages/common/src/lib/common/tools.ts (hunk 19 above)
Comment (foxriver76) [reply packages/common/src/lib/common/tools.ts:1294]: yes, but not necessarily in this PR.

**Review by AlCalzone — COMMENTED** [review]
Sorry, got some more :)

File: packages/common/src/lib/common/tools.ts (hunk 21 above)
Comment (foxriver76) [reply packages/common/src/lib/common/tools.ts:1375]: currently only used in main.js and main.js can put into it what it wants, just needs to be an Object so we can handle the data type
